### DISCUSSION TITLE
Are k_msleep(), k_sleep() etc guaranteed to sleep for at least the time specified?

### DISCUSSION THREAD

**Discussion by cmorganBE** [zephyrproject-rtos/zephyr#72751] [Q&A]
I understand they can be interrupted and will return the remaining time to sleep, but the question is if they aren't interrupted via k_wakeup(), will k_msleep(1), or k_msleep(5) etc sleep for at least 1ms and at least 5ms?

Ran into an issue with FreeRTOS vTaskDelay() where there is no guarantee that the sleep ticks is a full tick, so you can sleep for 5 ticks via vTaskDelay(5) and end up sleeping for only ~4 ticks. This is a significant limitation and wanted to understand if Zephyr was in the same spot or if there were techniques used by the OS to provide a "sleep for at least X" promise.

The documentation doesn't indicate one way or the other afaict.

**Comment by aescolar** [comment]
CC @andyross

**Reply by cmorganBE** [reply]
@andyross?

**Comment by andyross** [comment]
Yes, k_sleep() and variants will internally round up to the next tick boundary after the desired expiration and are guaranteed to sleep at least that long (or to return a non-zero result if they couldn't).  Though be aware getting reliable intervals with this kind of code is dangerous even with that promise.  Look at absolute timeouts, sys_timepoint, and k_timer's periodic interrupt facility for a better choice.
